### PR TITLE
[Hybrid Planning] Improve action cancellation

### PR DESCRIPTION
**Pull request by AndyZe**
Action cancellation wasn't working well, so...

- Move long-running callbacks into new threads
- Make a unique callback group for each action server
- Add a script to test cancellation

Test with:

`ros2 launch moveit_hybrid_planning hybrid_planning_demo.launch.py`

`ros2 run moveit_hybrid_planning cancel_action`

This work is sponsored by RE2 Robotics.

### CODE REVIEW

File: moveit_ros/hybrid_planning/global_planner/global_planner_component/src/global_planner_component.cpp
Comment (tylerjw) [reply moveit_ros/hybrid_planning/global_planner/global_planner_component/src/global_planner_component.cpp:90]: I wouldn't copy that from ros2... that seems about as bad as their spin_until_future_complete implementation.

https://stackoverflow.com/a/58222149

Doing this is technically creating a memory violation because the thread object might get cleaned up after this scope exits and its state is no longer valid.

Until we get `jthread` from c++20 which can detach a thread without creating a memory violation, we should instead always maintain the lifetime of thread object for as long as it runs and join it before cleaning up the memory.

Comment (AndyZe) [reply moveit_ros/hybrid_planning/global_planner/global_planner_component/src/global_planner_component.cpp:90]: any suggestion on how to actually do that? I could see something like storing a vector of `std::threads`, and joining all of them in the destructor. Sigh

Comment (AndyZe) [reply moveit_ros/hybrid_planning/global_planner/global_planner_component/src/global_planner_component.cpp]: OK, I added the std::thread as class member so there's no concern about scope. I also added a simple 30-line example of action server cancellation to have more confidence that it works. LGTM at this point.

Comment (henningkayser) [reply moveit_ros/hybrid_planning/global_planner/global_planner_component/src/global_planner_component.cpp]: Thank you! Two small things before merging this. You can drop the unique_ptr here. Simply make the thread a member, the default constructor will leave it uninitialized. That saves you the nullptr logic. Instead, before joining, threads should be checked if they are joinable, otherwise it allows for an error condition https://en.cppreference.com/w/cpp/thread/thread/join

File: moveit_ros/hybrid_planning/global_planner/global_planner_component/src/global_planner_component.cpp
Comment (AndyZe) [reply moveit_ros/hybrid_planning/global_planner/global_planner_component/src/global_planner_component.cpp]: I don't want to make this more complicated than it already is. I wish `join` had a timeout but it doesn't. I'm sure @tylerjw has a strong opinion   ;)

Comment (henningkayser) [reply moveit_ros/hybrid_planning/global_planner/global_planner_component/src/global_planner_component.cpp]: I think @zflat is right, that this wouldn't solve thread synchronization. Indeed, rejecting the goal would be a valid solution since we only support one thread at a time. Also, it wouldn't have to be joined in that case, the action would have to be canceled first. For this we wouldn't even need a lock guard, you could simply check if the thread is active (=joinable).

Comment (AndyZe) [reply moveit_ros/hybrid_planning/global_planner/global_planner_component/src/global_planner_component.cpp]: > rejecting the goal would be a valid solution since we only support one thread at a time

Well, that doesn't really make sense with the way RE2 is using this. We need to be able to accept new goals in a rapid-fire way.

Comment (AndyZe) [reply moveit_ros/hybrid_planning/global_planner/global_planner_component/src/global_planner_component.cpp]: How about I put a timeout around `join` like this. At least we will be able to throw a nice, human-readable error if thread joining takes too long.  (From my testing, it doesn't seem to be a problem in reality.)

https://stackoverflow.com/a/33017819

```
// Terminate the thread.
auto future = std::async(std::launch::async, &std::thread::join, &thread);
if (future.wait_for(std::chrono::seconds(5)) 
    == std::future_status::timeout) {
  /* --- Do something, if thread has not terminated within 5 s. --- */
}
```

Comment (AndyZe) [reply moveit_ros/hybrid_planning/global_planner/global_planner_component/src/global_planner_component.cpp]: https://user-images.githubusercontent.com/11284393/171268837-c72669b7-b1e0-4a30-8cef-2fedf99ced97.mp4

**Comment by AndyZe** [comment]
Can I get a final review? @henningkayser @JafarAbdi

**Review by henningkayser — APPROVED** [review]
I'm not a huge fan of the timeout, but I agree that this is probably the most practical approach for now. It's strictly an improvement to the old implementation, and in practice, I don't think that it even matters a lot in these situations.

**Pull request merged**
